### PR TITLE
HLRC: Add activate watch action

### PR DESCRIPTION
**Pull request by iverase**
Adds activate watch action to the high level rest client.

Relates #29827

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchRequest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            : Optional.empty();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return builder;
+    }
+
+    @Override
+    public String toString() {
```
Comment (pcsanwald) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchRequest.java]: do we need to override `toString` here? seems relatively harmless, but wasn't sure if it's necessary.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherIT.java [hunk 2]
```diff
@@ -72,4 +77,31 @@ public void testDeleteWatch() throws Exception {
...
+
+            ActivateWatchResponse activateWatchResponse2 = highLevelClient().watcher().activateWatch(new ActivateWatchRequest(watchId),
+                RequestOptions.DEFAULT);
+            assertThat(activateWatchResponse2.getStatus().state().isActive(), is(true));
+            assertThat(activateWatchResponse1.getStatus().state().getTimestamp(),
+                lessThan(activateWatchResponse2.getStatus().state().getTimestamp()));
+
+        }
+
+        // Activate watch that does not exist
```
Comment (pcsanwald) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherIT.java]: I'd split this into a separate test case, `testActivateWatchThatExists` and `testActivateWatchThatDoesNotExist`

File: client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherRequestConvertersTests.java [hunk 3]
```diff
@@ -75,4 +76,14 @@ public void testDeleteWatch() {
         assertEquals("/_xpack/watcher/watch/" + watchId, request.getEndpoint());
         assertThat(request.getEntity(), nullValue());
     }
+
+    public void testActivateWatch() {
```
Comment (pcsanwald) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherRequestConvertersTests.java]: I think `testActivateWatchRequestConversion` is a clearer name.

**Review by pcsanwald — APPROVED** [review]
A few small naming suggestions and one test suggestion, but I don't want to block merging on re-review.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherIT.java (hunk 2 above)
Comment (hub-cap) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherIT.java]: +1

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/WatcherDocumentationIT.java [hunk 4]
```diff
@@ -132,4 +135,59 @@ public void onFailure(Exception e) {
         }
     }
 
+    public void testActivateWatch() throws Exception {
```
Comment (hub-cap) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/WatcherDocumentationIT.java]: We will have a new ay to deal with the docs+test, we should wait to merge until it is merged.

**Review by hub-cap — CHANGES_REQUESTED** [review]
Wow im super sorry, i had done a review and forgot to click submit :/

Update to today, and we will very soon have a better way to test responses (you do not have a response test) and to generate docs, in the next day or two. We should hold off until then to merge. Feel free to fix up my nits tho.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/WatcherDocumentationIT.java (hunk 4 above)
Comment (iverase) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/WatcherDocumentationIT.java]: @hub-cap which is the issue that it is introducing this change?

Comment (hub-cap) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/WatcherDocumentationIT.java]: #34125 and it looks like its merged!

**Comment by iverase** [comment]
@hub-cap Added response test and changed the docs. I think it is getting closer.

**Comment by hub-cap** [comment]
Yea we backport all of the HLRC APIs to the current 6.x. Some of them landed in 6.4, most will land in 6.5

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchRequest.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import org.elasticsearch.client.Validatable;
+import org.elasticsearch.protocol.xpack.watcher.PutWatchRequest;
+
+import java.io.Closeable;
+import java.util.Objects;
+
+/**
+ * A request to explicitly activate a watch.
+ */
+public final class ActivateWatchRequest implements Validatable, Closeable {
```
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchRequest.java]: Any reason this is Closeable?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchRequest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    private final String watchId;
+
+    public ActivateWatchRequest(String watchId) {
+        if (watchId == null) {
+            throw new IllegalArgumentException("Watch identifier is required");
```
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchRequest.java]: there is a `Objects.requireNonNull` helper you can use here

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchResponse.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        ActivateWatchResponse that = (ActivateWatchResponse) o;
+        return Objects.equals(status, that.status);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(status);
+    }
+
+    private static final ParseField STATUS_FIELD = new ParseField("status");
```
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchResponse.java]: This is typically on teh top of these classes.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchResponse.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    private static ConstructingObjectParser<ActivateWatchResponse, Void> PARSER =
+        new ConstructingObjectParser<>("activate_watch_response", true,
+            a -> new ActivateWatchResponse((WatchStatus) a[0]));
+
+    static {
```
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchResponse.java]: this too is on top

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchRequest.java (hunk 5 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActivateWatchRequest.java]: You may also have a look at `TimedRequest`, assuming this needs the timeouts (I dont know if it does or not, you would have to research the transport layer action)

**Review by hub-cap — APPROVED** [review]
minor nits, then :shipit:

**Comment by iverase** [comment]
@elasticmachine test this please

**Comment by iverase** [comment]
@elasticmachine test this please

**Comment by iverase** [comment]
@hub-cap The commit cannot be back ported as WatchStatus has not been back ported. If necessary I can coordinate with @jtibshirani to do so.

**Comment by jtibshirani** [comment]
Just backported it to 6.x in 83e30eeeb5a5c255bfd32f40faf60c62ca75e69b, sorry for the delay!

**Pull request merged**